### PR TITLE
Emcs nrs support

### PR DESCRIPTION
**Pull request by mauro-fanton**

### CODE REVIEW

File: app/uk/gov/hmrc/excisemovementcontrolsystemapi/controllers/DraftExciseMovementController.scala [hunk 1]
```diff
@@ -53,8 +55,28 @@ class DraftExciseMovementController @Inject()(
 
   private def handleSuccess(implicit request: ValidatedXmlRequest[NodeSeq]): Future[Result] = {
 
+    val newMovement: Movement = createMovementFomMessage(request)
+    val saveMovementFn = movementMessageService.saveMovementMessage(newMovement)
+    val submitToNrsFn = submitToNrsService(newMovement)
+
+    //todo: Here we do not do anything id the NRS fails. THe NRS confluence page says we can do that
+    // handle the error if we want. I think the common approach by all the service is not doing anything.
+    // Also is this the right place to submit to the NRS?.
```
Comment (ChrisCookOC) [inline app/uk/gov/hmrc/excisemovementcontrolsystemapi/controllers/DraftExciseMovementController.scala:64]: The right place might be the Submit Connector? Just thinking that the messages we get in from the SubmitController (IE818 etc) will also need to be sent to NRS?

File: app/uk/gov/hmrc/excisemovementcontrolsystemapi/services/NrsService.scala [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+      metaData = NrsMetadata.create(payload, emcsUtils,notableEventId, identityData,dateTimeService.nowUtc.toString,
+        userAuthToken, userHeaderData, exciseNumber)
+      encodedPayload = emcsUtils.encode(payload)
+      nrsPayload = NrsPayload(encodedPayload, metaData)
+      retrievedNrsResponse <- nrsConnector.sendToNrs(nrsPayload)
+    } yield retrievedNrsResponse)
+      .recover {
+        case NonFatal(e) =>
+          //Todo: Catching the error and not throwing here for the moment. But do we want to throw here?
+          logger.error(s"[NrsService] - Error when submitting to Non repudiation system (NRS) with message: ${e.getMessage}", e)
```
Comment (ChrisCookOC) [inline app/uk/gov/hmrc/excisemovementcontrolsystemapi/services/NrsService.scala:64]: I guess an argument that we might not want to throw this is its NonFatal.

File: app/uk/gov/hmrc/excisemovementcontrolsystemapi/services/NrsService.scala [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+    val payload = request.body.toString
+    val userHeaderData = request.headersAsMap
+    val message = request.message
+    val exciseNumber = ernsMapper.getSingleErnFromMessage(message, request.validErns)
+    val notableEventId = nrsEventIdMapper.mapMessageToEventId(message)
+
+    (for {
+      identityData <- retrieveIdentityData()
+      userAuthToken = retrieveUserAuthToken(headerCarrier)
```
Comment (mpbostock) [inline app/uk/gov/hmrc/excisemovementcontrolsystemapi/services/NrsService.scala:59]: We don't retrieve the API Key from the user's token do we? Don't we just need to add the Authorization token to the NRS call similar to how we do it for EIS?

**Review by mpbostock — CHANGES_REQUESTED** [review]
There are a lot of changes that seem to be from other branches on this

File: app/uk/gov/hmrc/excisemovementcontrolsystemapi/connectors/NrsConnector.scala [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    appConfig: AppConfig,
+    emcsUtils: EmcsUtils,
+    metrics: Metrics
+)(implicit val ec: ExecutionContext, val futures: Futures) extends Retrying with Logging
+{
+
+  def sendToNrs(payload: NrsPayload)(implicit hc: HeaderCarrier): Future[Either[Int, NonRepudiationSubmissionAccepted]] = {
+
+    val timer = metrics.defaultRegistry.timer("emcs.nrs.submission.timer").time()
+    val correlationId = emcsUtils.generateCorrelationId
```
[the remark this answers is not in the document]
Comment (mauro-fanton) [reply app/uk/gov/hmrc/excisemovementcontrolsystemapi/connectors/NrsConnector.scala:47]: yes, this should be the same correlation of the submission. Looking into it

File: app/uk/gov/hmrc/excisemovementcontrolsystemapi/config/AppConfig.scala [hunk 5]
```diff
@@ -24,11 +24,14 @@ import scala.concurrent.duration.{DAYS, Duration, FiniteDuration, HOURS, MINUTES
...
 class AppConfig @Inject()(config: Configuration, servicesConfig: ServicesConfig) {
-
   val appName: String = config.get[String]("appName")
 
   lazy val eisHost: String = servicesConfig.baseUrl("eis")
+  lazy val nrsHost:  String = servicesConfig.baseUrl("nrs")
+
   lazy val systemApplication: String = config.get[String]("system.application")
+  lazy val nrsAuthorisationToken: String = servicesConfig.getString("microservice.services.nrs.api-key")
+  lazy val nrsRetries: Seq[FiniteDuration] = config.get[Seq[FiniteDuration]]("nrs.retries")
```
[the remark this answers is not in the document]
Comment (mauro-fanton) [reply app/uk/gov/hmrc/excisemovementcontrolsystemapi/config/AppConfig.scala:34]: done

File: app/uk/gov/hmrc/excisemovementcontrolsystemapi/connectors/NrsConnector.scala [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    }
+      .map { response: HttpResponse =>
+      response.status match {
+        case ACCEPTED =>
+          val submissionId = response.json.as[NonRepudiationSubmissionAccepted]
+          logger.info(s"[NrsConnector] - Non repudiation submission accepted with nrSubmissionId: ${submissionId.nrSubmissionId}")
+          Right(submissionId)
+        case _ =>
+          //todo: Add explicit audit error
+          logger.error(s"[NrsConnector] - Error when submitting to Non repudiation system (NRS) with status: ${response.status}, correlationId: $correlationId")
```
[the remark this answers is not in the document]
Comment (mauro-fanton) [reply app/uk/gov/hmrc/excisemovementcontrolsystemapi/connectors/NrsConnector.scala:61]: done

File: app/uk/gov/hmrc/excisemovementcontrolsystemapi/services/NrsService.scala [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+      metaData = NrsMetadata.create(payload, emcsUtils,notableEventId, identityData, dateTimeService.nowUtc.toString,
+        userAuthToken, userHeaderData, exciseNumber)
+      encodedPayload = emcsUtils.encode(payload)
+      nrsPayload = NrsPayload(encodedPayload, metaData)
+      retrievedNrsResponse <- nrsConnector.sendToNrs(nrsPayload)
+    } yield retrievedNrsResponse)
+      .recover {
+        case NonFatal(e) =>
+          //Todo: Catching the error and not throwing here for the moment. But do we want to throw here?
+          logger.error(s"[NrsService] - Error when submitting to Non repudiation system (NRS) with message: ${e.getMessage}", e)
```
[the remark this answers is not in the document]
Comment (mauro-fanton) [reply app/uk/gov/hmrc/excisemovementcontrolsystemapi/services/NrsService.scala:69]: done

File: app/uk/gov/hmrc/excisemovementcontrolsystemapi/utils/Retrying.scala [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+              futures.delay(delay).flatMap {_ => loop(attemptNumber + 1, tail)}
+            } else Future.fromTry(result)
+        }
+
+      block(attemptNumber).transformWith { result: Try[HttpResponse] =>
+        result match {
+          case Success(response) if Status.isSuccessful(response.status) => Future.successful(response)
+          case s: Success[HttpResponse] => retryIfPossible(s)
+          case f @ Failure(ex) =>
+             logger.error(s"[Retrying] - EMCS_API_RETRY error when retrying: url $url with message ${ex.getMessage}", ex)
```
[the remark this answers is not in the document]
Comment (mauro-fanton) [reply app/uk/gov/hmrc/excisemovementcontrolsystemapi/utils/Retrying.scala:57]: done

File: conf/application.conf [hunk 9]
```diff
@@ -108,20 +108,29 @@ scheduler {
...
+        port = 9389
+    }
   }
 }
 
 # The application system that submitted the message (previously known as originating system)
 system.application = "uk.gov.hmrc.cdio.cbit.emcs.core.api"
+
+# Defines the number and individual backoff delays for NRS submission retries
+nrs.retries = ["1s", "2s", "4s"]
```
[the remark this answers is not in the document]
Comment (mauro-fanton) [reply conf/application.conf:136]: done

File: test/uk/gov/hmrc/excisemovementcontrolsystemapi/connector/EISSubmissionConnectorSpec.scala [hunk 10]
```diff
@@ -97,7 +100,9 @@ class EISSubmissionConnectorSpec extends PlaySpec with BeforeAndAfterEach with E
     when(metrics.defaultRegistry.timer(any).time()) thenReturn timerContext
     when(ie815Message.messageType).thenReturn("IE815")
     when(ie815Message.consignorId).thenReturn("123")
-    when(emcsUtils.getSingleErnFromMessage(any, any)).thenReturn("123")
+    //todo: dalate
+    //when(emcsUtils.getSingleErnFromMessage(any, any)).thenReturn("123")
```
[the remark this answers is not in the document]
Comment (mauro-fanton) [reply test/uk/gov/hmrc/excisemovementcontrolsystemapi/connector/EISSubmissionConnectorSpec.scala:104]: done

File: app/uk/gov/hmrc/excisemovementcontrolsystemapi/services/NrsService.scala (hunk 3 above)
Comment (mauro-fanton) [reply app/uk/gov/hmrc/excisemovementcontrolsystemapi/services/NrsService.scala:60]: actually I am reviewing this

Comment (mauro-fanton) [reply app/uk/gov/hmrc/excisemovementcontrolsystemapi/services/NrsService.scala:60]: yes, that is different from the API-Key which we need to get from NRS

File: app/uk/gov/hmrc/excisemovementcontrolsystemapi/models/eis/EISErrorMessage.scala [hunk 11]
```diff
@@ -23,8 +23,9 @@ object EISErrorMessage {
              ern: String,
              message: String,
              correlationId: String,
-             messageTypes: String
+             messageTypes: String,
   ): String = {
+    //todo: Sore messageIdentifier from the header of the message as meesageId
```
[the remark this answers is not in the document]
Comment (mauro-fanton) [reply app/uk/gov/hmrc/excisemovementcontrolsystemapi/models/eis/EISErrorMessage.scala:28]: removed the todo

**Review by ChrisCookOC — DISMISSED** [review]

**Review by ChrisCookOC — APPROVED** [review]

File: app/uk/gov/hmrc/excisemovementcontrolsystemapi/config/AppConfig.scala [hunk 12]
```diff
@@ -24,10 +24,13 @@ import scala.concurrent.duration.{DAYS, Duration, FiniteDuration, HOURS, MINUTES
 
 @Singleton
 class AppConfig @Inject()(config: Configuration, servicesConfig: ServicesConfig) {
-
   val appName: String = config.get[String]("appName")
 
   lazy val eisHost: String = servicesConfig.baseUrl("eis")
+  lazy val nrsHost:  String = servicesConfig.baseUrl("nrs")
+
+  lazy val nrsAuthorisationToken: String = servicesConfig.getString("microservice.services.nrs.api-key")
```
Comment (mpbostock) [inline app/uk/gov/hmrc/excisemovementcontrolsystemapi/config/AppConfig.scala:32]: You can get a string from the servicesConfig without putting the `microservice.services` root folder at the start and just using `nrs.api-key`

File: conf/application.conf
Comment (mpbostock) [inline conf/application.conf:1]: This file now has inconsistent spacing, some lines have 4 spaces and some have 2. Can we make it consistent please?

File: app/uk/gov/hmrc/excisemovementcontrolsystemapi/services/NrsService.scala [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+      metaData = NrsMetadata.create(payload, emcsUtils,notableEventId, identityData, dateTimeService.nowUtc.toString,
+        userAuthToken, userHeaderData, exciseNumber)
+      encodedPayload = emcsUtils.encode(payload)
+      nrsPayload = NrsPayload(encodedPayload, metaData)
+      retrievedNrsResponse <- nrsConnector.sendToNrs(nrsPayload, correlationId)
+    } yield retrievedNrsResponse)
+      .recover {
+        case NonFatal(e) =>
+          logger.warn(s"[NrsService] - Error when submitting to Non repudiation system (NRS) with message: ${e.getMessage}", e)
+          Left(INTERNAL_SERVER_ERROR)
```
Comment (mpbostock) [inline app/uk/gov/hmrc/excisemovementcontrolsystemapi/services/NrsService.scala:70]: The failure on this call i.e. the Left(Int) response is not used apart from the tests, should we just refactor this to just return a NonRepudiationSubmissionAccepted?

Comment (mauro-fanton) [reply app/uk/gov/hmrc/excisemovementcontrolsystemapi/services/NrsService.scala]: done

File: app/uk/gov/hmrc/excisemovementcontrolsystemapi/config/AppConfig.scala (hunk 12 above)
Comment (mauro-fanton) [reply app/uk/gov/hmrc/excisemovementcontrolsystemapi/config/AppConfig.scala]: done

**Review by mpbostock — APPROVED** [review]

**Review by mpbostock — APPROVED** [review]

**Pull request merged**